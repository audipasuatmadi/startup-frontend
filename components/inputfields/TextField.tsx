import React, { useState, useRef } from 'react';

export interface TextFieldProps {
  name: string;
  id: string;
  type: string;
  value?: string;
  placeholder?: string;
  disabled?: boolean;

  labelText?: string;

  helperText?: string;
  error?: boolean;
  solid?: boolean;

  className?: string;

  onChange?: (
    e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => void;

  onKeyPress?: (e: React.KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>) => void;

  leadingIcon?: JSX.Element;
  trailingIcon?: JSX.Element;
}


const getIfDisabledStyle = (disabled: boolean) => {
  return `${
    disabled ? 'text-gray-400 cursor-not-allowed' : 'text-gray-600 cursor-text'
  }`;
};

const getIfSolid = (solid: boolean, error: boolean | undefined) => {
  return `${solid ? 'bg-gray-200' : `bg-none border-2`} ${
    error ? 'border-red-500 border-2' : 'border-gray-300'
  }`;
};

const getFocusedStyle = (solid: boolean) => {
  return `${
    solid
      ? `focus-within:ring-4 ring-brand ring-opacity-50`
      : `focus-within:border-brand focus-within:ring-4 ring-brand ring-opacity-50`
  }`;
};

const getTextFieldContainerStyle = (
  disabled: boolean,
  error: boolean | undefined,
  solid: boolean
) => {
  const finalStyle = `rounded ${getIfDisabledStyle(disabled)} ${getIfSolid(
    solid,
    error
  )} ${disabled ? '' : getFocusedStyle(solid)}`;
  return finalStyle;
};

const TextField = ({
  id,
  labelText,
  name,
  type,
  disabled,
  error,
  helperText,
  leadingIcon,
  onChange,
  onKeyPress,
  placeholder,
  solid,
  value,

  trailingIcon,

  className,
}: TextFieldProps) => {
  const [isFocused, setIsFocused] = useState(false);
  const myRef = useRef<HTMLInputElement>(null);

  return (
    <React.Fragment>
      {labelText && (
        <>
          <label
            className={`ml-5${error ? ' text-red-600' : ' text-gray-400'} ${
              isFocused ? 'text-brand' : ''
            }`}
            htmlFor={id}>
            {labelText}
          </label>
          <br />
        </>
      )}

      <div
        className={`inline-block px-5 py-2 ${
          labelText ? 'mt-1' : ''
        } ${getTextFieldContainerStyle(disabled!, error, solid!)}${className ? ` ${className}` : ''}`}
        onClick={() => myRef.current && myRef.current.focus()}>
        <div className='flex items-center'>
          {leadingIcon}

          <input
            id={id}
            name={name}
            type={type}
            placeholder={placeholder}
            disabled={disabled}
            ref={myRef}
            className={`bg-transparent focus:outline-none w-full${
              leadingIcon ? ' ml-2' : ''
            }${trailingIcon ? ' mr-2' : ''}`}
            onChange={disabled ? undefined : onChange}
            onKeyPress={disabled ? undefined : onKeyPress}
            onFocus={() => setIsFocused(true)}
            onBlur={() => setIsFocused(false)}
            value={value}
          />

          {trailingIcon}
        </div>
      </div>

          {helperText && (
            <>
              <label
                className={`block ml-5${error ? ' text-red-600' : ' text-gray-400'}`}
                htmlFor={id}>
                {helperText}
              </label>
            </>
          )}
    </React.Fragment>
  );
};

export default TextField;
