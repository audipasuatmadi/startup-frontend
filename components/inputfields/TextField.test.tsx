import React from 'react'
import TextField, { TextFieldProps } from './TextField';
import { ShallowWrapper, shallow } from 'enzyme';

const dummyText = 'txt';

const baseProps: TextFieldProps = {
  id: dummyText,
  labelText: dummyText,
  name: dummyText,
  type: 'text',
  value: '',
};

describe('textfield functionality tests', () => {
  let wrapper: ShallowWrapper;
  let props: TextFieldProps;

  const renderComponent = (props: TextFieldProps) => <TextField {...props} />;
  afterEach(() => {
    wrapper.unmount();
    props = baseProps;
  });

  it('is able to change values & sync with the particular source', () => {
    let myVal = 'hey';

    const handleChangeVal = (
      e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
    ) => {
      myVal = e.target.value;
    };

    props = { ...baseProps, value: myVal, onChange: handleChangeVal };
    wrapper = shallow(renderComponent(props));
    wrapper.find('input').simulate('change', { target: { value: dummyText } });

    expect(myVal).toMatch(dummyText);
  });

  it('is not able to change values & sync with the particular source if disabled', () => {
    let myVal = 'hey';

    const handleChangeVal = (
      e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
    ) => {
      myVal = e.target.value;
    };

    props = {
      ...baseProps,
      value: myVal,
      onChange: handleChangeVal,
      disabled: true,
    };
    wrapper = shallow(renderComponent(props));
    wrapper.find('input').simulate('change', { target: { value: dummyText } });

    expect(wrapper.find('input').name()).toBe('input');
    expect(myVal).not.toMatch(dummyText);
  });
});

describe('textfield rendering tests', () => {
  let wrapper: ShallowWrapper;
  let props: TextFieldProps;

  const renderComponent = (props: TextFieldProps) => <TextField {...props} />;
  afterEach(() => {
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
    props = baseProps;
  });

  it('renders correctly with base props', () => {
    wrapper = shallow(renderComponent(baseProps));
  });

  it('renders correctly in the solid state', () => {
    props = { ...baseProps, solid: true };
    wrapper = shallow(renderComponent(props));
  });

  it('renders correctly with helper text', () => {
    props = { ...baseProps, helperText: dummyText };
    wrapper = shallow(renderComponent(props));
  });

  it('renders correctly in the error state', () => {
    props = { ...baseProps, error: true };
    wrapper = shallow(renderComponent(props));
  });

  it('renders correctly in the disabled state', () => {
    props = { ...baseProps, disabled: true };
    wrapper = shallow(renderComponent(props));
  });
});
