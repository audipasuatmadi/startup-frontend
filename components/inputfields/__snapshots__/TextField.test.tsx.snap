// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`textfield rendering tests renders correctly in the disabled state 1`] = `
<Fragment>
  <label
    className="ml-5 text-gray-400 "
    htmlFor="txt"
  >
    txt
  </label>
  <br />
  <div
    className="inline-block px-5 py-2 mt-1 rounded text-gray-400 cursor-not-allowed bg-none border-2 border-gray-300 "
    onClick={[Function]}
  >
    <div
      className="flex items-center"
    >
      <input
        className="bg-transparent focus:outline-none w-full"
        disabled={true}
        id="txt"
        name="txt"
        onBlur={[Function]}
        onFocus={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
</Fragment>
`;

exports[`textfield rendering tests renders correctly in the error state 1`] = `
<Fragment>
  <label
    className="ml-5 text-red-600 "
    htmlFor="txt"
  >
    txt
  </label>
  <br />
  <div
    className="inline-block px-5 py-2 mt-1 rounded text-gray-600 cursor-text bg-none border-2 border-red-500 border-2 focus-within:border-brand focus-within:ring-4 ring-brand ring-opacity-50"
    onClick={[Function]}
  >
    <div
      className="flex items-center"
    >
      <input
        className="bg-transparent focus:outline-none w-full"
        id="txt"
        name="txt"
        onBlur={[Function]}
        onFocus={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
</Fragment>
`;

exports[`textfield rendering tests renders correctly in the solid state 1`] = `
<Fragment>
  <label
    className="ml-5 text-gray-400 "
    htmlFor="txt"
  >
    txt
  </label>
  <br />
  <div
    className="inline-block px-5 py-2 mt-1 rounded text-gray-600 cursor-text bg-gray-200 border-gray-300 focus-within:ring-4 ring-brand ring-opacity-50"
    onClick={[Function]}
  >
    <div
      className="flex items-center"
    >
      <input
        className="bg-transparent focus:outline-none w-full"
        id="txt"
        name="txt"
        onBlur={[Function]}
        onFocus={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
</Fragment>
`;

exports[`textfield rendering tests renders correctly with base props 1`] = `
<Fragment>
  <label
    className="ml-5 text-gray-400 "
    htmlFor="txt"
  >
    txt
  </label>
  <br />
  <div
    className="inline-block px-5 py-2 mt-1 rounded text-gray-600 cursor-text bg-none border-2 border-gray-300 focus-within:border-brand focus-within:ring-4 ring-brand ring-opacity-50"
    onClick={[Function]}
  >
    <div
      className="flex items-center"
    >
      <input
        className="bg-transparent focus:outline-none w-full"
        id="txt"
        name="txt"
        onBlur={[Function]}
        onFocus={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
</Fragment>
`;

exports[`textfield rendering tests renders correctly with helper text 1`] = `
<Fragment>
  <label
    className="ml-5 text-gray-400 "
    htmlFor="txt"
  >
    txt
  </label>
  <br />
  <div
    className="inline-block px-5 py-2 mt-1 rounded text-gray-600 cursor-text bg-none border-2 border-gray-300 focus-within:border-brand focus-within:ring-4 ring-brand ring-opacity-50"
    onClick={[Function]}
  >
    <div
      className="flex items-center"
    >
      <input
        className="bg-transparent focus:outline-none w-full"
        id="txt"
        name="txt"
        onBlur={[Function]}
        onFocus={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
  <label
    className="block ml-5 text-gray-400"
    htmlFor="txt"
  >
    txt
  </label>
</Fragment>
`;
